internals/overlord/logstate/loki/loki_test.go

```diff
@@ -0,0 +1,241 @@
+    {
+      "stream": {
+        "pebble_service": "svc1"
+      },
+      "values": [
+          [ "1704026090000000000", "log line #1" ],
+          [ "1704026092000000000", "log line #3" ],
+          [ "1704026094000000000", "log line #5" ],
+          [ "1704026097000000000", "log line #8" ]
+      ]
+    },
+    {
+      "stream": {
+        "pebble_service": "svc2"
+      },
+      "values": [
+          [ "1704026091000000000", "log line #2" ],
+          [ "1704026096000000000", "log line #7" ]
+      ]
+    },
+    {
+      "stream": {
+        "pebble_service": "svc3"
+      },
+      "values": [
+          [ "1704026093000000000", "log line #4" ],
+          [ "1704026095000000000", "log line #6" ]
+      ]
+    },
+    {
+      "stream": {
+        "pebble_service": "svc4"
+      },
+      "values": [
+          [ "1704026098000000000", "log line #9" ]
+      ]
+    }
+  ]
+}
+`
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		c.Assert(r.Method, Equals, http.MethodPost)
+		c.Assert(r.Header.Get("Content-Type"), Equals, "application/json")
+
+		reqBody, err := io.ReadAll(r.Body)
+		c.Assert(err, IsNil)
+		expFlattened, err := flattenJSON(expected)
+		c.Assert(err, IsNil)
+		c.Assert(string(reqBody), Equals, expFlattened)
+	}))
+	defer server.Close()
+
+	client := loki.NewClient(&plan.LogTarget{Location: server.URL})
+	for _, entry := range input {
+		err := client.Write(context.Background(), entry)
+		c.Assert(err, IsNil)
+	}
+
+	err := client.Flush(context.Background())
+	c.Assert(err, IsNil)
+}
+
+func (*suite) TestFlushCancelContext(c *C) {
+	serverCtx, killServer := context.WithCancel(context.Background())
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		select {
+		case <-serverCtx.Done():
+		// Simulate a slow-responding server
+		case <-time.After(10 * time.Second):
+		}
+	}))
+	defer server.Close()
+	defer killServer()
+
+	client := loki.NewClient(&plan.LogTarget{Location: server.URL})
+	err := client.Write(context.Background(), servicelog.Entry{
+		Time:    time.Now(),
+		Service: "svc1",
+		Message: "this is a log line\n",
+	})
+	c.Assert(err, IsNil)
+
+	flushReturned := make(chan struct{})
+	go func() {
+		// Cancel the Flush context quickly
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Microsecond)
+		defer cancel()
+
+		err := client.Flush(ctx)
+		c.Assert(errors.Is(err, clienterr.Timeout{}), Equals, true)
+		close(flushReturned)
+	}()
+
+	// Check Flush returns quickly after context timeout
+	select {
+	case <-flushReturned:
+	case <-time.After(1 * time.Second):
+		c.Fatal("lokiClient.Flush took too long to return after context timeout")
+	}
+}
+
+func (*suite) TestServerTimeout(c *C) {
+	restore := loki.SetRequestTimeout(1 * time.Microsecond)
+	defer restore()
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(2 * time.Microsecond)
```

Comment:
I think this should be a long-ish time like `time.Second` in case there are sequencing issues? I'm not sure that can happen, but just in case. And in the success case the test will still run fast.

Reply:
I'm not sure what you mean by "sequencing issues"

I wouldn't want it taking a whole second in the failure case

Reply:
Something like:

```go
stopRequest := make(chan struct{})
defer func() { close(stopRequest) }
server := ... {
    <-stopRequest
}
```